### PR TITLE
Add script to setup external cloud provider

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

File: templates/external_cloud_setup.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+
+if [ "$#" -ne 1 ]; then
+    echo "One param for cluster-name needed"
+    echo "The usage: external_cloud_setup.sh <cluster-name>"
+    exit -1
```
Comment (seanschneeweiss) [inline templates/external_cloud_setup.sh:21]: `-1` is out of range according to https://tldp.org/LDP/abs/html/exitcodes.html#EXITCODESREF
```suggestion
    exit 1
```

File: templates/external_cloud_setup.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+# Copyright 2020 The Kubernetes Authors.
```
Comment (seanschneeweiss) [inline templates/external_cloud_setup.sh:2]:
```suggestion
# Copyright 2022 The Kubernetes Authors.
```

File: templates/external_cloud_setup.sh [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+CLUSTER_NAME=$1
+
+echo "##########################################"
+echo "Create kubeconfig file ${CLUSTER_NAME}.kubeconfig for cluster ${CLUSTER_NAME}"
+clusterctl get kubeconfig ${CLUSTER_NAME} --namespace default > ./${CLUSTER_NAME}.kubeconfig
+
+echo "##########################################"
+echo "Create secret cloud-config"
+./create_cloud_conf.sh clouds.yaml openstack > /tmp/cloud.conf
```
Comment (seanschneeweiss) [inline templates/external_cloud_setup.sh:46]: What do you think about using `mktemp -d` for creating a temporary directory where the `cloud.conf` can be saved to? And afterwards you can remove that directory.

File: templates/external_cloud_setup.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+echo "##########################################"
+echo "Apply OCCM manifests"
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/master/manifests/controller-manager/cloud-controller-manager-roles.yaml
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/master/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml
+
+echo "##########################################"
+echo "Install CNI, currently it's calico"
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://docs.projectcalico.org/v3.23/manifests/calico.yaml
```
Comment (mdbooth) [inline templates/external_cloud_setup.sh:58]: We don't do this for other templates. Should we include this?

File: templates/external_cloud_setup.sh (hunk 3 above)
Comment (jichenjc) [reply templates/external_cloud_setup.sh:46]: yes , will do that

File: templates/external_cloud_setup.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+echo "Create secret cloud-config"
+./create_cloud_conf.sh clouds.yaml openstack > /tmp/cloud.conf
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig create secret -n kube-system generic cloud-config --from-file=/tmp/cloud.conf
+rm /tmp/cloud.conf
+
+echo "##########################################"
+echo "Apply OCCM manifests"
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/master/manifests/controller-manager/cloud-controller-manager-roles.yaml
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/master/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml
+kubectl --kubeconfig=./${CLUSTER_NAME}.kubeconfig apply -f https://raw.githubusercontent.com/kubernetes/cloud-provider-openstack/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml
```
[the remark this answers is not in the document]
Comment (jichenjc) [reply templates/external_cloud_setup.sh:54]: right, will make it a version specific

File: templates/external_cloud_setup.sh (hunk 4 above)
Comment (jichenjc) [reply templates/external_cloud_setup.sh:58]: I am considering whether or not do this ,

maybe provide a flag instead ..

**Comment by mnaser** [comment]
For the record, we use ClusterResourceSet to work around this

**Pull request closed without merging**